### PR TITLE
Improve build system

### PR DESCRIPTION
**Pull request by Robzz**
Here are a bunch of changes related to libSGM's build system that I use on a fork, I'm opening a PR here in case you'd want to integrate them:

* Make building samples optional (samples depend on GLFW, but the main lib doesn't, so it's now possible to build without errors without installing GLFW)
* Add ability to build as a shared library, in which case some symbols are exported
* Add install target to the build system
* Add a CMake FindModule, so we can find the library using CMake's `find_package` command.

### CODE REVIEW

**Comment by ykitta-fixstars** [comment]
Thank you for your contribution.
I start to check

**Comment by Robzz** [comment]
Hey, just to let you know I added a commit introducing CUDA 9 support since I kinda need that :)  
By the way, are there any changes you'd like me to do ?

**Comment by ykitta-fixstars** [comment]
I'm so sorry for late reply.
I checked building library with following parameters:

- CUDA 8.0, build with `-DENABLE_SAMPLES=on -DLIBSGM_SHARED=on`
- CUDA 8.0, build with `-DENABLE_SAMPLES=on -DLIBSGM_SHARED=off`
- CUDA 9.0, build with `-DENABLE_SAMPLES=on -DLIBSGM_SHARED=on`
- CUDA 9.0, build with `-DENABLE_SAMPLES=on -DLIBSGM_SHARED=off`
- CUDA 9.0, build with `-DENABLE_SAMPLES=off -DLIBSGM_SHARED=on`
- CUDA 9.0, build with `-DENABLE_SAMPLES=off -DLIBSGM_SHARED=off`

(For all parameters, OS is Ubuntu 16.04 and CMake version is 3.10.1.)

I checked fail to build 1 case -- CUDA 9.0, build with `-DENABLE_SAMPLES=on -DLIBSGM_SHARED=off`.
(Other case succeed to build.)

Here is compilation log:

```
[ 68%] Linking CXX executable stereo_test
../../src/libsgm.a(stereo_sgm.cpp.o): In function `sgm::CudaStereoSGMResources::CudaStereoSGMResources(int, int, int, int, int, sgm::EXECUTE_INOUT)':
stereo_sgm.cpp:(.text._ZN3sgm22CudaStereoSGMResourcesC2EiiiiiNS_13EXECUTE_INOUTE[_ZN3sgm22CudaStereoSGMResourcesC5EiiiiiNS_13EXECUTE_INOUTE]+0x31d): undefined reference to `nppiFilterMedianGetBufferSize_16u_C1R'
../../src/libsgm.a(sgm_generated_median_filter.cu.o): In function `sgm::details::median_filter(unsigned short const*, unsigned short*, void*, int, int)':
/tmp/libSGM/src/median_filter.cu:29: undefined reference to `nppiFilterMedian_16u_C1R'
collect2: error: ld returned 1 exit status
```

Add to link `${CUDA_nppif_LIBRARY}` of sample programs CMakeLists.txt.

**Comment by Robzz** [comment]
Nice catch :) I pushed the fix.

**Comment by ykitta-fixstars** [comment]
To check working FindLibSGM.cmake correctly, I tried to build a simple program. (OS is Ubuntu 16.04)

```cmake
cmake_minimum_required(VERSION 2.8)
project(sample)
find_package(OpenCV)
find_package(LibSGM)

include_directories(${LIBSGM_INCLUDE_DIRS})
add_executable(sample main.cpp)
target_link_libraries(sample ${LIBSGM_LIBRARIES} ${OpenCV_LIBS})
```

main.cpp:
```c++
#include <libsgm.h>

int main()
{
    cv::Mat left, right;
    // Try to call libSGM library
    sgm::StereoSGM ssgm(left.cols, left.rows, 64, 8, 8, sgm::EXECUTE_INOUT_HOST2HOST);
}
```

But it is failed:

```sh
developer@9c13f2d1eaa6:/tmp/sample/build$ find /usr/local -name FindLibSGM.cmake
/usr/local/FindLibSGM.cmake
developer@9c13f2d1eaa6:/tmp/sample/build$ cmake ..
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenCV: /usr/local (found version "3.4.0")
CMake Warning at CMakeLists.txt:4 (find_package):
  By not providing "FindLibSGM.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "LibSGM", but
  CMake did not find one.

  Could not find a package configuration file provided by "LibSGM" with any
  of the following names:

    LibSGMConfig.cmake
    libsgm-config.cmake

  Add the installation prefix of "LibSGM" to CMAKE_PREFIX_PATH or set
  "LibSGM_DIR" to a directory containing one of the above files.  If "LibSGM"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/sample/build
```

I'm not familiar with FindXXX.cmake of  CMake, how do I fix it to work?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,12 @@ include_directories(../include)
 
 find_package(CUDA REQUIRED)
 
+if(CUDA_VERSION VERSION_GREATER_EQUAL "9.0")
```
Comment (ykitta-fixstars) [inline src/CMakeLists.txt]: `VERSION_GREATER_EQUAL` statement is enabled to use CMake 3.7 or newer.
Replace with VERSION_GREATER and VERSION_EQUAL for older versions.
https://cmake.org/cmake/help/v3.7/release/3.7.html#commands

**Comment by Robzz** [comment]
Basically when calling CMake `find_package(XXX)`, CMake looks for a file called `FindXXX.cmake` or `XXXConfig.cmake` in one of the directories contained in `CMAKE_MODULE_PATH`. When you install libSGM, chances are the resulting path to `FindLibSGM.cmake` is not included in your `CMAKE_MODULE_PATH`. One way to resolve the issue is to add the directory containing `FindLibSGM.cmake` to `CMAKE_MODULE_PATH`.  
But a better way is the following: when CMake can't find a library through `find_package`, it creates a cache variable called `XXX_DIR` that will default to some invalid value (`XXX-NOTFOUND` if I remember correctly). You can use this variable to tell CMake where the findmodule is located.  
For instance say you install libSGM by setting your install prefix to `/usr/local`, you'll end up with `FindLibSGM.cmake` at `/usr/local/FindLibSGM.cmake`. Set the variable `libSGM_DIR` to `/usr/local` and CMake should then be able to find libSGM.

File: src/CMakeLists.txt (hunk 1 above)
Comment (Robzz) [reply src/CMakeLists.txt]: Should be fixed now

**Comment by Robzz** [comment]
FYI I added another commit to rename the configuration header, as `config.h` is way too generic and is going to cause hard to debug issues.

File: src/CMakeLists.txt (hunk 1 above)
Comment (ykitta-fixstars) [reply src/CMakeLists.txt]: Thank you!

**Comment by ykitta-fixstars** [comment]
I could to build a simple program with `-DCMAK_MODULE_PATH`. Thank you!
(Do cmake `-DlibSGM_DIR` is failed. Somehow that parameter search `XXXconfig.cmake` only.)

Thank you for contribution! I'll merge this.

**Pull request merged**